EXILED/Exiled.Events/Patches/Events/Scp939/PlayingFootstep.cs

```diff
@@ -0,0 +1,72 @@
+// -----------------------------------------------------------------------
+// <copyright file="PlayingFootstep.cs" company="Exiled Team">
+// Copyright (c) Exiled Team. All rights reserved.
+// Licensed under the CC BY-SA 3.0 license.
+// </copyright>
+// -----------------------------------------------------------------------
+
+using Exiled.API.Features;
+
+namespace Exiled.Events.Patches.Events.Scp939
+{
+    using System.Collections.Generic;
+    using System.Reflection.Emit;
+
+    using Exiled.API.Features.Pools;
+    using Exiled.Events.EventArgs.Scp939;
+    using Exiled.Events.Handlers;
+
+    using HarmonyLib;
+    using PlayerRoles.FirstPersonControl.Thirdperson;
+    using PlayerRoles.PlayableScps.Scp939.Ripples;
+
+    using static HarmonyLib.AccessTools;
+
+    /// <summary>
+    /// Patches <see cref="FootstepRippleTrigger.OnFootstepPlayed(AnimatedCharacterModel, float)" />
+    /// to add the <see cref="Scp939.PlayingFootstep" /> event.
+    /// </summary>
+    [HarmonyPatch(typeof(FootstepRippleTrigger), nameof(FootstepRippleTrigger.OnFootstepPlayed))]
+    internal static class PlayingFootstep
+    {
+        private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstruction> instructions, ILGenerator generator)
+        {
+            List<CodeInstruction> newInstructions = ListPool<CodeInstruction>.Pool.Get(instructions);
+
+            Label ret = generator.DefineLabel();
+
+            int offset = 2;
+            int index = newInstructions.FindLastIndex(i => i.opcode == OpCodes.Call && i.Calls(Method(typeof(RippleTriggerBase), nameof(RippleTriggerBase.CheckVisibility)))) + offset;
+
+            newInstructions.InsertRange(index, new CodeInstruction[]
+            {
+                // Player player = Player.Get(hub);
+                new CodeInstruction(OpCodes.Ldarg_1).MoveLabelsFrom(newInstructions[index]),
+                new(OpCodes.Callvirt, PropertyGetter(typeof(CharacterModel), nameof(CharacterModel.OwnerHub))),
+                new(OpCodes.Call, Method(typeof(API.Features.Player), nameof(API.Features.Player.Get), new[] { typeof(ReferenceHub) })),
+
+                // true;
+                new(OpCodes.Ldc_I4_1),
+
+                // PlayingFootstepEventArgs ev = (player, true);
+                new(OpCodes.Newobj, GetDeclaredConstructors(typeof(PlayingFootstepEventArgs))[0]),
+                new(OpCodes.Dup),
+
+                // Handlers.Scp939.OnPlayingFootstep(ev);
+                new(OpCodes.Call, Method(typeof(Scp939), nameof(Scp939.OnPlayingFootstep))),
+
+                // if (!IsAllowed)
+                //      return;
+                new(OpCodes.Callvirt, PropertyGetter(typeof(PlayingFootstepEventArgs), nameof(PlayingFootstepEventArgs.IsAllowed))),
+                new(OpCodes.Brfalse_S, ret),
+            });
+
+            newInstructions[newInstructions.Count - 1].WithLabels(ret);

```

Comment:
not a serious thing, but it is better to use `labels.Add` in situations where you dont need `WithLabels` return
```suggestion
            newInstructions[newInstructions.Count - 1].labels.Add(ret);

```